x-pack/plugins/task_manager/server/task_running/task_runner.ts

```diff
@@ -369,59 +363,72 @@ export class TaskManagerRunner implements TaskRunner {
     }
   }
 
-  private validateTaskParams({ taskInstance }: RunContext) {
-    let error;
+  private async validateTask({ taskInstance }: RunContext) {
     const { state, taskType, params, id, numSkippedRuns = 0 } = taskInstance;
+
+    if (!this.requeueInvalidTasksConfig.enabled) {
+      return { state };
+    }
+
     const { max_attempts: maxAttempts } = this.requeueInvalidTasksConfig;
+    const { paramsSchema, indirectParamsSchema, latestTypeVersion } = this.definition;
+    let hasValidationError = false;
 
-    try {
-      const paramsSchema = this.definition.paramsSchema;
-      if (paramsSchema) {
+    // validate task params
+    if (paramsSchema) {
+      try {
         paramsSchema.validate(params);
-      }
-    } catch (err) {
-      this.logger.warn(`Task (${taskType}/${id}) has a validation error: ${err.message}`);
-      if (numSkippedRuns < maxAttempts) {
-        error = createSkipError(err);
-      } else {
-        this.logger.warn(
-          `Task Manager has reached the max skip attempts for task ${taskType}/${id}`
-        );
+      } catch (err) {
+        hasValidationError = true;
+        this.logger.warn(`Task (${taskType}/${id}) has a validation error: ${err.message}`);
+        if (numSkippedRuns < maxAttempts) {
+          return { error: createSkipError(err), state };
+        }
       }
     }
 
-    return { ...(error ? { error } : {}), state };
-  }
+    if (this.task?.loadIndirectParams) {
+      const { data } = await this.task.loadIndirectParams();
+      const indirectParams = data?.indirectParams;
+      const typeVersion = data?.typeVersion;
 
-  private async validateIndirectTaskParams({ taskInstance }: RunContext) {
-    let error;
-    const { state, taskType, id, numSkippedRuns = 0 } = taskInstance;
-    const { max_attempts: maxAttempts } = this.requeueInvalidTasksConfig;
-    const indirectParamsSchema = this.definition.indirectParamsSchema;
+      // validate runtime version
+      if (typeVersion && latestTypeVersion && typeVersion > latestTypeVersion) {
+        hasValidationError = true;
+        this.logger.warn(
+          `Task (${taskType}/${id}) has a newer version(${typeVersion}) than expected((${latestTypeVersion}))`
+        );
+        if (numSkippedRuns < maxAttempts) {
+          return {
+            error: createSkipError(
+              new Error('Task has already been run by a newer Kibana version')
+            ),
+            state,
+          };
+        }
+      }
 
-    if (this.task?.loadIndirectParams && !!indirectParamsSchema) {
-      const { data } = await this.task.loadIndirectParams();
-      if (data) {
+      // validate indirect params e.g. connector or rule.params
+      if (indirectParamsSchema && indirectParams) {
```

Comment:
Just to validate my understanding:  even if the type version is newer than Kibana expects, the first few times trying to run this task (up to `maxAttempts`), we'll go ahead and try to validate the params anyway, and if successful, we'll run the task.  That appears to be what the code is doing, and my understanding of how we wanted it to work.

There's something about this in the code that looks a bit funny, but I think it will work fine, so will probably need a comment in the final PR.

I think we'd aalways want to change the log warn above ("has a newer version") to be debug, since it will just be confusing to see this if the rule actually DOES get run.